Add transition GUI to the E+ Auxiliary CLI

Myoldmopar opened the pull request:
Pull request overview
---------------------

Allows launching the Python based version updater tool from the E+ CLI:




https://github.com/user-attachments/assets/f73a1273-6dcf-48ea-91cd-48ae82da8891


I'll kick off a test package so that any interested party could try it out.  As of now I've only exercised the E+ CLI on Linux, but it works the same as the EP Launch CLI, which is happy on all platforms.  Also, I had to make some improvements to the updater tool, and it's tested heavily across platforms itself, so I think it should be OK.....

Myoldmopar committed:
Add transition GUI to the E+ Auxiliary CLI

Myoldmopar commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -979,7 +979,7 @@ if(LINK_WITH_PYTHON)
       POST_BUILD # TODO: I don't think we want to quote the generator expression
       COMMAND ${CMAKE_COMMAND} -E env --unset=PIP_REQUIRE_VIRTUALENV
-      ${Python_EXECUTABLE} -m pip install --target="$<TARGET_FILE_DIR:energyplusapi>/python_lib" --upgrade energyplus-launch==3.7.2
+      ${Python_EXECUTABLE} -m pip install --target="$<TARGET_FILE_DIR:energyplusapi>/python_lib" --upgrade energyplus-launch==3.7.2 energyplus-transition-tools==2.1.1
```
Just grab the transition tool package, at version 2.1.1 which I just released this morning after some cleanups.

jmarrec replied:
When do we start using https://github.com/Myoldmopar/EnergyPlusPythonApps/ as a meta package itself?

jmarrec replied:
* https://github.com/Myoldmopar/EnergyPlusPythonApps/pull/6

Myoldmopar commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -278,6 +278,50 @@ sys.argv.append("energyplus")
+
+            cmd += R"python(
+from energyplus_transition.runner import main_gui
+main_gui(True)
```
If the auxiliary updater keywords are passed to the E+ CLI, it will launch the transition GUI and leave.  This could be modified later to like pass the IDF file in, but I truly don't think that's necessary.  This is launching a GUI, no one should be integrated this auxiliary argument into some sort of automated workflow.

Myoldmopar replied:
And yes, there is quite a bit of duplicate from the EP Launch code right above.  Would be nice to collect things into the PythonEngine class and out of here.  But doesn't have to be right now.

jmarrec replied:
> If the auxiliary updater keywords are passed to the E+ CLI, it will launch the transition GUI and leave. This could be modified later to like pass the IDF file in, but I truly don't think that's necessary. This is launching a GUI, no one should be integrated this auxiliary argument into some sort of automated workflow.

It'd sure be nice if one could bypass the GUI and do something like `energyplus auxiliary updater myfile.idf` (potentially with a `--up-to-version 24.2.0`) to directly transition the file in the command line.

Myoldmopar replied:
Yes, that would be great, let's do that soon.

Myoldmopar reviewed (commented):
Note - this will still execute the underlying Fortran based transition binaries.  It just launches the Python GUI.  IDFVersionUpdater is I think broken again on Ubuntu 24.04, so this would eliminate the need for that.  I am open to removing it entirely in favor of this.

jmarrec commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -278,6 +278,50 @@ sys.argv.append("energyplus")
+        });
+
+        auto *updaterSubCommand = auxiliaryToolsSubcommand->add_subcommand("updater", "IDF Version Updater");
+        updaterSubCommand->add_option("args", python_fwd_args, "Extra Arguments forwarded to IDF Version Updater")->option_text("ARG ...");
```
What extra args does this accept? I'm confused.

Myoldmopar replied:
It doesn't...for now.  This could be considered future proofing and not hurt anything right now.

mjwitte commented:
Tested this on Windows. 

1. Transitioned a v9.1 file up to 25.1, worked as expected, except the new files are in the "run directory" which is the \PreProcess\IDFVersionUpdater folder. Can we have an option to save the files in the location of the original idf?
2. After launching the GUI, the command prompt doesn't return until the GUI is closed. Is that the expected behavior?
![image](https://github.com/user-attachments/assets/e58df049-de78-40f6-b9ed-b80802a12eaf)
3. The current IDFVersionUpdater has an option to process a list of files.
4. Can we add a double-clickable shortcut to launch this from a folder (so I don't have to remember the syntax to type)

Myoldmopar committed:
Pin transition tools to 2.1.2, add a shortcut

Myoldmopar commented:
@mjwitte Thank you!! OK, got nearly everything resolved for you.

1. Added a menu option that defaults to True which will copy the final version of the transitioned file back to the original input file dir
2. Yes, that's fine
3. I added in the ability to do lst files.  It's a bit hacky, but seems to work really well.  Would love your testing.
4. Yes, added a shortcut
5. I did not do this, but am open to it.  I just can't remember what all files I need to grab.

I am going to kick off another test package with all this in it.  Let me know if you find anything else!

jmarrec commented:
> Added a menu option that defaults to True which will copy the final version of the transitioned file back to the original input file dir

That makes me puzzled. There is no reason it would not be in the original dir, is there? If I run this

```python
transition_exe = Path('/path/to/EnergyPlus/PreProcess/IDFVersionUpdater/Transition-VXXX-to-VYYY')
subprocess.check_call([str(transition_exe), '/full/path/to/file.idf'], cwd=transition_exe.parent)
```

Then the idf object is transition at its original location to begin with

mjwitte commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -995,6 +995,7 @@ if(BUILD_PACKAGE)
       # NOTE: The install command COMPONENTS should line up so that they are run at the same packaging step
       install(CODE "execute_process(COMMAND powershell.exe -ExecutionPolicy Bypass -File ${PROJECT_SOURCE_DIR}/scripts/dev/create_shortcut.ps1 -TargetPath \"$<TARGET_FILE:energyplus>\" -ShortcutPath \"$<TARGET_FILE_DIR:energyplus>/EPLaunchPython.lnk\" -Arguments \"auxiliary eplaunch\")" COMPONENT Auxiliary)
+      install(CODE "execute_process(COMMAND powershell.exe -ExecutionPolicy Bypass -File ${PROJECT_SOURCE_DIR}/scripts/dev/create_shortcut.ps1 -TargetPath \"$<TARGET_FILE:energyplus>\" -ShortcutPath \"$<TARGET_FILE_DIR:energyplus>/IDFVersionUpdater.lnk\" -Arguments \"auxiliary updater\")" COMPONENT Auxiliary)
       install(FILES $<TARGET_FILE_DIR:energyplus>/EPLaunchPython.lnk DESTINATION "./" COMPONENT Auxiliary)
```
The IDFVersionUpdater shortcut is not there. Does it need a line like this one?

Myoldmopar replied:
Oh sheesh.  Yes.  I've got another round of tweaks on the updater tool itself that I'm wrapping up and will tag/pin a new version, and update this here.

Myoldmopar committed:
Bump updater version, add idf version updater shortcut to install

Myoldmopar committed:
Added a macro to generate a thin app for Macs [decent_ci_skip] [skip ci]

Myoldmopar committed:
Some more tweaks [decent_ci_skip] [skip ci]

Myoldmopar committed:
Some more tweaks [decent_ci_skip] [skip ci]

Myoldmopar committed:
Some more tweaks [decent_ci_skip] [skip ci]

Myoldmopar committed:
Some more tweaks [decent_ci_skip] [skip ci]

Myoldmopar committed:
Some more tweaks [decent_ci_skip] [skip ci]

Myoldmopar committed:
Some more tweaks [decent_ci_skip] [skip ci]

Myoldmopar committed:
Some more tweaks [decent_ci_skip] [skip ci]

Myoldmopar committed:
Some more tweaks [decent_ci_skip] [skip ci]

Myoldmopar committed:
Fix typo in launcher script, move back to just apple for app bundles [decent_ci_skip]

Myoldmopar commented on cmake/Shortcut.plist.in:
I think I'll plan on just naming this info.plist.in since it becomes info.plist.

Myoldmopar committed:
Trying to fix signage...@jmarrec? [decent_ci_skip]

Myoldmopar commented:
So there is a [ValueError](https://github.com/NREL/EnergyPlus/actions/runs/13689514711/job/38280050905#step:14:201) saying that launcher.sh isn't signed.  The launcher.sh is a minimal little bash script that goes inside the newly generated .app bundles for EPLaunch and VersionUpdater.  I am not sure that those need to be signed, but I found the spot where other bash scripts (runenergyplus) were being signed, so I added it to that list.  I'm not 100% sure how declaring the name there is enough, but it seems to work for runenergyplus and the others.  Anxious to learn about how to address this properly since there will be more coming later. 

@jmarrec FYI and HELP! :laughing:

jmarrec commented on cmake/Install.cmake:
```diff
@@ -678,6 +678,7 @@ if(APPLE)
       "runenergyplus"
       "runepmacro"
       "runreadvars"
+      "launcher.sh"
```
I think you need to sign the .app bundle itself.

jmarrec replied:
I don't think bash scripts really need to be signed? 

In any case, your launcher.sh If I'm reading this PR correctly does NOT live at INSTALL_PATH/launcher.sh but at INSTALL_PATH/AppName.app/Contents/MacOS/launcher.sh, right?

Myoldmopar committed:
Update CMakeLists.txt [decent_ci_skip]

OK, so the install(signer script could not find the launcher.sh file. It looks like it is trying to find it in the packages/Unspecified/data folder.  But I had used COMPONENT Auxiliary in the install( for the app bundle.  I am kinda hopeful that maybe this will work.  By not specifying a COMPONENT, it will end up in the Unspecified package to match the signer script.  Let's see.

jmarrec committed:
Copy the thin apps from the Products dir

jmarrec committed:
Rename the launcher.sh like the APPNAME actually, that's the way it works

jmarrec committed:
Update plist

jmarrec committed:
Add icon + display name

jmarrec committed:
Just install the presigned apps, remove the complex on-the-fly creation

jmarrec committed:
Entitlements + osascript for debug

Myoldmopar committed:
Merge changes together, disable apple apps for now

jmarrec committed:
Better script with error handling.

jmarrec committed:
Rerereresign

Myoldmopar committed:
Merge pull request #10976 from NREL/PackageIDFUpdater2

#10975 - Add mac launchers for PythonApps

Myoldmopar committed:
Merge in @jmarrec changes

Myoldmopar committed:
Intermediate commit, just need to defer finding desktop template

Myoldmopar committed:
Tweaking .desktop file generation

Myoldmopar commented:
Aaaaaaaahhhhh  @jmarrec !!!!!

https://github.com/user-attachments/assets/3554d796-4bc3-4740-b6dc-c1ff241381fa

Myoldmopar committed:
Trying something on windows [decent_ci_skip] [skip ci]

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PackageIDFUpdater

Myoldmopar committed:
Move pip command to standalone Python file, check before trying to pip install each time

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PackageIDFUpdater

Myoldmopar committed:
Trying a different windows launcher script [decent_ci_skip] [actions skip]

Myoldmopar committed:
Fixed up the windows_gui_launcher script [decent_ci_skip] [actions skip]

Myoldmopar committed:
Packaging tcl tk on Linux now seems to be in place

Myoldmopar committed:
Clang format

Myoldmopar committed:
Fix missing license and exclude cc in byref checker

Myoldmopar committed:
Fixed windows shortcut, leaving a temporary debug message for Linux [decent_ci_skip] [actions skip]

Myoldmopar committed:
And changed root shortcuts to use the windows gui launcher [decent_ci_skip]

Myoldmopar commented:
This is super close now.  I'm glad CI is all clean, and all the package builds were successful.  What I see now:

- On Windows, the start menu shortcut was one long quoted string instead of a quoted executable and argument, so I had to tweak the QtIFW to handle that better.  Should be good to go.
- On Windows, I also build some clickable shortcuts in the E+ install root called EP-LaunchPython and IDFVersionUpdater.  They weren't behaving, so they are fixed now.
- Outside of maybe some renaming, and documentation, Windows should be good to go.
- On Linux, I'm still not able to get it to launch without setting LD_LIBRARY_PATH from outside the call to EnergyPlus.  Even though I'm setting it inside E+ as I make the call to the Python interpreter.  I added a debug message to help diagnose, and will kick off test 9.

Myoldmopar committed:
Fix typo in shortcut name [decent_ci_skip]

Myoldmopar committed:
Fix root shortcut arguments, windows is good

Myoldmopar committed:
Fix library path issue on Linux, should work now, removing debug

Myoldmopar committed:
And fixup linux desktop file as well

jmarrec commented on cmake/energyplus.desktop.in:
```diff
@@ -2,7 +2,7 @@
 Name=EnergyPlus ENERGYPLUS_VERSION
 Comment=Run EnergyPlus Launch
-Exec=ENERGYPLUS_INSTALL_DIR/energyplus auxiliary eplaunch
+Exec=env LD_LIBRARY_PATH=ENERGYPLUS_INSTALL_DIR/python_lib/lib-dynload ENERGYPLUS_INSTALL_DIR/energyplus auxiliary eplaunch
```
I think you're perhaps going about it the wrong way.

First, on mac, what we do is to:
* Copy the tcl libs
* Use `install_name_tool` to change the `_tkinter.so` so look for these libs and NOT at `/opt/homebrew/xxxx` . See cmake/PythonFixUpTclTk.py
* That way no need to mess with the LD_LIBRARY_PATH or anything

So at the very least, you should probably do the same with Linux.

```diff
(py312)(3.2.2)julien@EnergyPlus-build-release$ ldd Products/python_lib/lib-dynload/_tkinter.cpython-312-x86_64-linux-gnu.so
	linux-vdso.so.1 (0x0000734351f28000)
+	libtk8.6.so => /lib/x86_64-linux-gnu/libtk8.6.so (0x0000734351d84000)
+	libtcl8.6.so => /lib/x86_64-linux-gnu/libtcl8.6.so (0x0000734351bd7000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x0000734351800000)
	libXft.so.2 => /lib/x86_64-linux-gnu/libXft.so.2 (0x0000734351bbc000)
	libfontconfig.so.1 => /lib/x86_64-linux-gnu/libfontconfig.so.1 (0x0000734351b6b000)
-	libX11.so.6 => /lib/x86_64-linux-gnu/libX11.so.6 (0x0000734351a2c000)
```

**But Honestly, I don't see why we should ship libtk/libtcl any more than libX11 for eg.**


On mac, there is a difference:

1) You must have brew installed to brew install tcl-tk . Ubuntu users have `apt`  already.
2) The newest brew recipe is incompatible, you have to specifically call `brew install tcl-tk@8`
     * cf https://formulae.brew.sh/formula/tcl-tk

FWIW, in OpenStudio when we build the .deb installer, we do set the dependencies that are missing in E+:

https://github.com/NREL/OpenStudio/blob/98207d21649408d964fbc2660110a4cbdb35eb7d/CMakeLists.txt#L1264

```cmake
set(CPACK_DEBIAN_PACKAGE_DEPENDS "libgomp1, libx11-6") # This is for EnergyPlus
```

jmarrec replied:
https://github.com/NREL/EnergyPlus/blob/4df1e4cdd70ef7094a0bf0f32a3a2d4ccbaca4fc/cmake/PythonFixUpTclTk.py#L117-L142

Myoldmopar replied:
One of my main goals is to continue to allow a non-sudo install of EnergyPlus on Linux. If the user doesn't want system symlinks, and wants to install it in a user-owned folder, they shouldn't need to follow up with sudo installing stuff.  I am not going to die on that hill, but I will risk being injured on it.  I have it working really well, so I think we should put 25.1 out there with this, and we can make a correction if needed as we go into the next version.

Myoldmopar committed:
Cleaning up after self-inspecting the changes

Myoldmopar commented on bin/PythonApps/EPLaunch.app/Contents/CodeResources:
OK, so I'll do a walkthrough now.  There are still a couple things left but it's very very close.

Myoldmopar replied:
Thanks to @jmarrec for getting these Python app bundles signed and packed up nicely.  Things are working amazingly on Mac.  I imagine we will want more of these as we add a few more apps (EpJSONEditor...)

Myoldmopar commented on cmake/Install.cmake:
```diff
@@ -766,6 +766,10 @@ cpack_ifw_configure_component(Symlinks SCRIPT cmake/qtifw/install_unix_createsym
 
+if (PYTHON_CLI)
+  cpack_ifw_configure_component(Unspecified SCRIPT cmake/qtifw/install_auxiliary_python_shortcuts.qs)
+endif()
```
Alright, so here is a new QtIFW script that is attached to the Unspecified component.  Could consider adding it to the Auxiliary or something if we wanted all the Python CLI app stuff in Auxiliary.  

Anyway, this script will handle any desktop integration steps to enhance the user experience.

Myoldmopar commented on cmake/PythonFixUpTclTk.py:
```diff
@@ -100,46 +101,80 @@ def get_linked_libraries(p: Path):
 if __name__ == "__main__":
 
-    if platform.system() != "Darwin":
+    if platform.system() != "Darwin" and platform.system() != 'Linux':
```
Now that we are packaging up the tcl and tk stuff for Ubuntu, we need to abort if we are not on one of those (so .... Windows :upside_down_face: )

Myoldmopar commented on cmake/PythonFixUpTclTk.py:
```diff
@@ -100,46 +101,80 @@ def get_linked_libraries(p: Path):
-
-    if any_missing:
-        sys.exit(1)
+    if platform.system() == 'Darwin':
```
If we are on Mac, we haven't changed anything from 24.2.  We still find the tk blobs, move things around, and run install_name_tool to get everyone happy.

Myoldmopar commented on cmake/PythonFixUpTclTk.py:
```diff
@@ -100,46 +101,80 @@ def get_linked_libraries(p: Path):
+            sys.exit(1)
+
+    elif platform.system() == 'Linux':
+        # On Linux, we need to copy the libtk.X.Y.so file from usually /usr/lib/x86_64-linux-gnu
```
On Linux, we are currently getting the libtk library, plus the tcl and tk config/data directories and bringing them along.  As it stands right now, on Linux you need to tell the process where to find libtk8.6.so in order to run the Python GUIs.  This is hidden from the user through the desktop integration and setting LD_LIBRARY_PATH.  But @jmarrec feels we should do a patchelf to fix up the _tkinter library instead.  I'm on board, and this will likely be fixed up before merge.  I will say, it does seem like everything is running happily though.

Myoldmopar replied:
Indeed the patchelf stuff worked!  Note that when you apt install python3-tk, you will get a _tkinter shared object library that is stripped of it's ELF section headers, and you cannot write an rpath into it.  A warning has been added to GHA if it encounters this condition.

Myoldmopar commented on cmake/PythonInstallCLITools.py:
```diff
@@ -0,0 +1,64 @@
+pkgs = {'energyplus_launch': '3.7.2', 'energyplus_transition_tools': '2.1.3'}
+if to_install := [f"{n}=={v}" for n, v in pkgs.items() if not (Path(argv[1]) / f"{n}-{v}.dist-info").exists()]:
+    print(f"PYTHON: CLI packages not found or out of date, pip installing these now: {to_install}")
+    run([executable, '-m', 'pip', 'install', f'--target={argv[1]}', '--upgrade', *to_install], check=True)
```
This is the tool that gets the pip installed tools dropped into the build tree.  Note it does a check on the installed version before doing the pip install to avoid constantly pip installing with each build if you have PYTHON_CLI on.

Myoldmopar commented on cmake/PythonSetupAPIinBuild.cmake:
```diff
@@ -1,13 +1,13 @@
 
 # informative messaging
-message("Setting up Python API, creating pyenergyplus package at ${EPLUS_EXE_DIR}/pyenergyplus")
+message("PYTHON: Setting up API, creating pyenergyplus package at ${EPLUS_EXE_DIR}/pyenergyplus")
```
Just changed wording to make it look a little nicer during the build.

Myoldmopar replied:
I would like to move these to regular CMake messages instead of print statements.

Myoldmopar commented on cmake/auxiliary.sh.in:
```diff
@@ -0,0 +1,5 @@
+trap '' HUP
+export LD_LIBRARY_PATH="$(dirname "$0")/python_lib/lib-dynload"
+"$(dirname "$0")/energyplus" auxiliary @ARGUMENT@ &
+trap - HUP
```
This is a sweet little bash script that allows execution of a script without popping open the terminal.  The shenanigans are top notch but work flawlessly.

Myoldmopar commented on cmake/qtifw/install_auxiliary_python_shortcuts.qs:
```diff
@@ -0,0 +1,29 @@
+      component.addOperation("CreateShortcut", "@TargetDir@/windows_gui_launcher.exe", target_dir + "/IDFVersionUpdater.lnk", "updater");
+    }
+  }
+}
```
This new QtIFW script will setup desktop integration for the Python CLI tools at install time.  On Linux it will create a .desktop file inside .local/share/applications, which allows the user to find EnergyPlus tools in the main activities view.  On Windows it will create shortcuts to the new tools in the start menu along with the rest of the E+ shortcuts.

Myoldmopar commented on cmake/qtifw/install_win_createstartmenu.qs:
```diff
@@ -47,11 +47,11 @@ function Component()
       component.addOperation("CreateShortcut", "@TargetDir@/ExampleFiles/ExampleFiles-ObjectsLink.html", target_dir + "/ExampleFiles Link to Objects.lnk");
       component.addOperation("CreateShortcut", "@TargetDir@/PreProcess/IDFEditor/IDFEditor.exe", target_dir + "/IDFEditor.lnk");
-      component.addOperation("CreateShortcut", "@TargetDir@/PreProcess/IDFVersionUpdater/IDFVersionUpdater.exe", target_dir + "/IDFVersionUpdater.lnk");
+      component.addOperation("CreateShortcut", "@TargetDir@/PreProcess/IDFVersionUpdater/IDFVersionUpdater.exe", target_dir + "/IDFVersionUpdater-Classic.lnk");
```
I'm renaming the other versions of EP Launch and Version Updater as "Classic" to start moving away from them.

Myoldmopar commented on scripts/dev/find_byref_bool_override.py:
```diff
@@ -96,7 +96,8 @@
                          'src/EnergyPlus/test_ep_as_library.cc',
-                         'EnergyPlusPgm.cc']
+                         'EnergyPlusPgm.cc',
+                         'src/EnergyPlus/WindowsGuiLauncher.cc']
```
A new source file that doesn't have a matching header, since it's just a tiny little exe.

Myoldmopar commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -979,8 +979,9 @@ if(LINK_WITH_PYTHON)
       POST_BUILD # TODO: I don't think we want to quote the generator expression
       COMMAND ${CMAKE_COMMAND} -E env --unset=PIP_REQUIRE_VIRTUALENV
-      ${Python_EXECUTABLE} -m pip install --target="$<TARGET_FILE_DIR:energyplusapi>/python_lib" --upgrade energyplus-launch==3.7.2
+      ${Python_EXECUTABLE} ${PROJECT_SOURCE_DIR}/cmake/PythonInstallCLITools.py "$<TARGET_FILE_DIR:energyplusapi>/python_lib"
```
Collect the pip install operations into a Python script to allow some extra intelligence.

Myoldmopar commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -991,11 +992,37 @@ if(BUILD_PACKAGE)
-      install(CODE "execute_process(COMMAND powershell.exe -ExecutionPolicy Bypass -File ${PROJECT_SOURCE_DIR}/scripts/dev/create_shortcut.ps1 -TargetPath \"$<TARGET_FILE:energyplus>\" -ShortcutPath \"$<TARGET_FILE_DIR:energyplus>/EPLaunchPython.lnk\" -Arguments \"auxiliary eplaunch\")" COMPONENT Auxiliary)
-      install(FILES $<TARGET_FILE_DIR:energyplus>/EPLaunchPython.lnk DESTINATION "./" COMPONENT Auxiliary)
+    if(PYTHON_CLI)
+      if(WIN32)
```
For Python CLI on Windows, we now build a tool called windows_gui_launcher.exe, whose role is to execute a command line tool without opening a terminal so that it can launch a GUI nicely.  We also create shortcuts in the EnergyPlus install root so that the user can find them there.   Perhaps I should move the IDFVersionUpdater into the PreProcess folder?

Myoldmopar commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -991,11 +992,37 @@ if(BUILD_PACKAGE)
+      elseif(APPLE)
+        # Pre-signed and notarized launcher apps
+        install(DIRECTORY "${PROJECT_SOURCE_DIR}/bin/PythonApps/EPLaunch.app" DESTINATION "." COMPONENT Auxiliary USE_SOURCE_PERMISSIONS)
+        install(DIRECTORY "${PROJECT_SOURCE_DIR}/bin/PythonApps/VersionUpdater.app" DESTINATION "." COMPONENT Auxiliary USE_SOURCE_PERMISSIONS)
```
On Apple, we just need to drop in the prepared app bundles.

Myoldmopar commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -991,11 +992,37 @@ if(BUILD_PACKAGE)
+        install(
+                FILES ${PROJECT_SOURCE_DIR}/cmake/energyplus.desktop.in
+                DESTINATION share/applications
+        )
```
On Linux we now need to create a couple shell scripts to launch EPLaunch and VersionUpdater right in the install root.  On second thought, maybe these aren't needed?  I'll consider removing them.

And we also now install a .desktop template inside install root / share/applications to be customized at install time by the QtIFW scripts.

Myoldmopar replied:
I decided to leave these little shortcuts in.  I honestly doubt many people will ever use them on Linux.  Either they will be happy on a terminal and just execute `./energyplus auxiliary eplaunch`, or they will use the desktop integration and click on the E+ icon on the dock.  But it's fine for now.

Myoldmopar commented on src/EnergyPlus/PythonEngine.cc:
```diff
@@ -361,6 +361,42 @@ namespace Python {
+        cmd += fmt::format("environ[\'TCL_LIBRARY\'] = \"{}/{}\"\n", sPathToPythonPackages, tclConfigDir);
+        cmd += fmt::format("environ[\'TK_LIBRARY\'] = \"{}/{}\"\n", sPathToPythonPackages, tkConfigDir);
+        return cmd;
+    }
```
I was duplicating this for each GUI, so I made a worker function out of it.

Myoldmopar commented on src/EnergyPlus/WindowsGuiLauncher.cc:
This is the worker function on Windows that can execute a script without opening a terminal.  The shenanigans are at a all time high, but this works so perfectly.  I'm quite happy with it.

Myoldmopar reviewed (commented):
Still a few things left to consider, but walkthrough is done.

- ~~Icons labeled and version numbered?~~
  - This isn't happening now.  I want to get a version of E+ out the door, and get the full experience across all platforms, and then make improvements based on that.  At a minimum, I suspect we'll end up making some new icons for EPLaunch vs VersionUpdater.  And we may need to tweak the Python code for those to get the icons to behave really nicely.  So this is for later.
- ~~add_custom_commands all need to identify their output files~~
  - I would still love this, but it needs a little work.  The BYPRODUCTS cannot be target-specific generators, and we do a lot of target-specific stuff in the add_custom_commands related to Python.  I definitely want to rethink that stuff but that's not today.
- [x] .desktop should be part of uninstall
- [ ] patchelf instead of LD_LIBRARY_PATH
- [x] remove IDFVersionUpdater classic?  At least from Linux?
- [ ] Update documentation
- [x] ~~Move IDFVersionUpdater shortcut into preprocess?~~
  - No, but I put a "VersionUpdater-NEWS.txt file in the PreProcess folder to help users find out about the new stuff 
- [x] ~~Remove the auxiliary.sh.in stuff on Linux, is it used?~~
  - For now I'm going to leave it because it allows a user to double click to launch the tools from the install folder.  If we can get the LD_LIBRARY_PATH thing solved, then it may not be as needed.  I will double check that we are creating similar names of these shortcuts in all platforms though.

Myoldmopar committed:
Add undo command to clean up the .desktop file during uninstall

Myoldmopar committed:
Remove IDFVersionUpdater XoJo version from Git and cpack install commands

Myoldmopar committed:
Apply shellcheck recommendation for export in auxiliary.sh.in

Myoldmopar committed:
Unified names of shortcuts in install folder

Myoldmopar committed:
Fixed up desktop file uninstall cleanup, added a NEWS file in the preprocess folder

jmarrec commented on src/EnergyPlus/WindowsGuiLauncher.cc:
```diff
@@ -0,0 +1,97 @@
+        return 1;
+    }
+
+    // Remove the executable name to get the directory
+    std::wstring directory(path);
+    size_t pos = directory.find_last_of(L"\\/");
+    if (pos != std::wstring::npos) {
+        directory = directory.substr(0, pos); // Extract just the directory path
+    }
+
+    // Construct the target application path (assuming it's in the same directory)
+    std::wstring targetApp = directory + L"\\energyplus.exe";
```
One could argue you should be using std::filesystem::path (which uses wchar_t on windows), instead of string manipulation. But you're already using Windows API, so screwed for screwed 🤷 

https://godbolt.org/z/G1Ydo6qEz

jmarrec replied:
I think you could call GetLastError to format the system error message. Untested:

```diff
diff --git a/src/EnergyPlus/WindowsGuiLauncher.cc b/src/EnergyPlus/WindowsGuiLauncher.cc
index 8d63d8b3f1..b8a007275a 100644
--- a/src/EnergyPlus/WindowsGuiLauncher.cc
+++ b/src/EnergyPlus/WindowsGuiLauncher.cc
@@ -45,30 +45,53 @@
 // OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 // POSSIBILITY OF SUCH DAMAGE.
 
+#include <filesystem>
 #include <string>
 #include <windows.h>
 
+void ErrorExit()
+{
+    // Retrieve the system error message for the last-error code
+
+    LPVOID lpMsgBuf;
+    DWORD dw = GetLastError();
+
+    if (FormatMessageW(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+                       NULL,
+                       dw,
+                       MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+                       (LPWSTR)&lpMsgBuf,
+                       0,
+                       NULL) == 0) {
+        MessageBoxW(NULL, TEXT("FormatMessage failed"), TEXT("Error"), MB_OK);
+        ExitProcess(dw);
+    }
+
+    MessageBoxW(NULL, (LPWSTR)lpMsgBuf, TEXT("Error"), MB_ICONERROR | MB_OK);
+
+    LocalFree(lpMsgBuf);
+    ExitProcess(dw);
+}
+
 int wWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPWSTR lpCmdLine, int nCmdShow)
 {
+    using namespace fs = std::filesystem;
+
     // Get the path of the current executable
-    wchar_t path[MAX_PATH];
-    if (GetModuleFileNameW(NULL, path, MAX_PATH) == 0) {
+    wchar_t c_path[MAX_PATH];
+    if (GetModuleFileNameW(NULL, c_path, MAX_PATH) == 0) {
         MessageBoxW(NULL, L"Failed to get module file name!", L"Error", MB_ICONERROR | MB_OK);
         return 1;
     }
 
     // Remove the executable name to get the directory
-    std::wstring directory(path);
-    size_t pos = directory.find_last_of(L"\\/");
-    if (pos != std::wstring::npos) {
-        directory = directory.substr(0, pos); // Extract just the directory path
-    }
+    fs::path thisExePath(c_path);
+    fs::path thisDirPath = thisExePath.parent_path();
 
     // Construct the target application path (assuming it's in the same directory)
-    std::wstring targetApp = directory + L"\\energyplus.exe";
+    fs::path targetApp = thisDirPath / L"energyplus.exe";
 
-    // Check if the file exists
-    if (GetFileAttributesW(targetApp.c_str()) == INVALID_FILE_ATTRIBUTES) {
+    if (!fs::is_regular_file(targetApp)) {
         MessageBoxW(NULL, L"Application not found!", L"Error", MB_ICONERROR | MB_OK);
         return 1;
     }
@@ -78,19 +101,19 @@ int wWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPWSTR lpCmdLine, int
     args += lpCmdLine;
 
     // Launch the target application
-    HINSTANCE hInstance2 = ShellExecuteW(NULL,              // Parent window
-                                         L"open",           // Operation
-                                         targetApp.c_str(), // Executable path
-                                         args.c_str(),      // Arguments
-                                         directory.c_str(), // Working directory
-                                         SW_HIDE            // Hide console
+    HINSTANCE hInstance2 = ShellExecuteW(NULL,                // Parent window
+                                         L"open",             // Operation
+                                         targetApp.c_str(),   // Executable path
+                                         args.c_str(),        // Arguments
+                                         thisDirPath.c_str(), // Working directory
+                                         SW_HIDE              // Hide console
     );
 
-    // Error handling
+    // Error handling: https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shellexecutew#return-value
     if ((intptr_t)hInstance2 <= 32) {
         std::wstring errorMsg = L"Failed to launch application! Error code: " + std::to_wstring((intptr_t)hInstance2);
         MessageBoxW(NULL, errorMsg.c_str(), L"Error", MB_ICONERROR | MB_OK);
-        return 1;
+        ErrorExit();
     }
 
     return 0;
```

Myoldmopar replied:
Deferring this until later, but very happy to get this tested and integrated later.

Myoldmopar committed:
Removing LD_LIBRARY_PATH workaround, and using patchelf instead.  If linux is happy then this just needs docs and merged.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PackageIDFUpdater

Myoldmopar commented:
And it is totally happy on Linux!  I just need to know from @jmarrec if those Windows changes are worth pulling in or not.  If not, then this could be ready to merge in.  I think now I'll just defer any documentation changes until my very last PrepFor251 style branch where I update the README and all the other silly stuff.

Myoldmopar commented:
OK, and it seems that the Windows changes are not important right now.  In the spirit of trying to get this in the hands of testers as soon as possible, I'm planning on merging this once CI finishes up, and tagging RC1.  I will await testing results and start modifying documentation for these changes in the meantime.

Myoldmopar commented on .github/workflows/release_linux.yml:
```diff
@@ -67,7 +67,7 @@ jobs:
         echo "Using Apt to install dependencies"
         sudo apt-get update
-        sudo apt-get install texlive texlive-xetex texlive-science libxkbcommon-x11-0 xorg-dev libgl1-mesa-dev
+        sudo apt-get install texlive texlive-xetex texlive-science libxkbcommon-x11-0 xorg-dev libgl1-mesa-dev patchelf
```
Patchelf is used to edit the ELF header in the _tkinter shared object library on Linux, similar to `install_name_tool` on Mac.

Myoldmopar commented on bin/IDFVersionUpdater/Run-Linux/IDFVersionUpdater:
Removed the IDFVersionUpdater for Linux....Windows and Mac are going to be gone with the next release.

Myoldmopar commented on cmake/energyplus.desktop.in:
```diff
@@ -0,0 +1,18 @@
+
+[Desktop Action VersionUpdater]
+Name=Open IDF Version Updater
+Exec=ENERGYPLUS_INSTALL_DIR/energyplus auxiliary updater
```
CLEAN little desktop file. :)

Myoldmopar commented on cmake/qtifw/install_auxiliary_python_shortcuts.qs:
```diff
@@ -0,0 +1,34 @@
+          "else echo 'Error: .desktop template file missing'; fi"
+        ],
+        "UNDOEXECUTE",
+        "rm", desktopFilePath
```
Even added an uninstall step to clean up the .desktop file path.

Myoldmopar commented on release/VersionUpdater-NEWS.txt:
```diff
@@ -0,0 +1,12 @@
+*Linux note*: if you try to launch the VersionUpdater program from the default files app, it
+may launch in a text editor (since it is a shell script). Just right click and ask to "Run as a Program".
+
+The classic version updater is included with this version on Windows and Mac, but will be removed in the future.
```
Users will find this file in the PreProcess folder if they go hunting for IDFVersionUpdater.  This is just one vector that I'll be using to alert the users of these changes.

Myoldmopar reviewed (commented):
OK, that's it.  Code walkthrough is done.  Merging this in.  If CI doesn't find anything in the normal `develop` testing, then I'll tag it as RC1 and start working on docs/release notes.
